### ISSUE TITLE
Crash in [@ neqo_common::codec::Encoder::encode_uint<T>]

### ISSUE THREAD

**Issue by mxinden** [mozilla/neqo#2132]
See [bugzilla#1919678](https://bugzilla.mozilla.org/show_bug.cgi?id=1919678).

Investigating ...

**Comment by mxinden** [comment]
Investigation thus far.

Panic occurs here:

https://github.com/mozilla/neqo/blob/d513712bc9d06e297af5cf317a9f6dd70e69a88b/neqo-common/src/codec.rs#L309

Within `Encoder::encode_uint`:

https://github.com/mozilla/neqo/blob/d513712bc9d06e297af5cf317a9f6dd70e69a88b/neqo-common/src/codec.rs#L302-L314

Called from `PacketBuilder::pn`:

https://github.com/mozilla/neqo/blob/d513712bc9d06e297af5cf317a9f6dd70e69a88b/neqo-transport/src/packet/mod.rs#L321-L354

Called by `Connection::add_packet_number`:

https://github.com/mozilla/neqo/blob/d513712bc9d06e297af5cf317a9f6dd70e69a88b/neqo-transport/src/connection/mod.rs#L1991-L2007

`n` can not be `>8` due to the line below where `MAX_PACKET_NUMBER_LEN` is `4`:

https://github.com/mozilla/neqo/blob/d513712bc9d06e297af5cf317a9f6dd70e69a88b/neqo-transport/src/packet/mod.rs#L343

In debug mode `n` can also not be `0` due to the line right after:

https://github.com/mozilla/neqo/blob/d513712bc9d06e297af5cf317a9f6dd70e69a88b/neqo-transport/src/packet/mod.rs#L344

But I assume in release mode it can.

Thus this `pn_len` calculation must somehow return `0`:

https://github.com/mozilla/neqo/blob/d513712bc9d06e297af5cf317a9f6dd70e69a88b/neqo-transport/src/connection/mod.rs#L2017-L2020

For that, `unacked_range` must be `0`.

In order for `unacked_range` to be `0`, `largest_acknowledged` must be equal to `pn`.

In other words, my suspicion is that we are trying to write the packet number of the currently largest acknowledged packet.

**Comment by mxinden** [comment]
Provoking the panic itself at the level of `Connection::add_packet_number()` by providing an equal `largest_acknowledged` and `next_pn` is easy:

``` rust
#[test]
fn add_packet_number() {
    test_fixture::fixture_init();
    let mut builder = PacketBuilder::short(neqo_common::Encoder::new(), false, None::<&[u8]>);
    let tx = crate::crypto::CryptoDxState::test_default();
    let largest_acknowledged = Some(tx.next_pn());
    let _ = Connection::add_packet_number(&mut builder, &tx, largest_acknowledged);
}
```

That said, I am unable to come up with a scenario where a `largest_acknowledged` equal to `next_pn` would ever be passed to `Connection::add_packet_number()`.

I will think about this some more.

**Comment by larseggert** [comment]
I wonder if this is related to this change:
https://github.com/mozilla/neqo/blob/d513712bc9d06e297af5cf317a9f6dd70e69a88b/neqo-transport/src/tracking.rs#L502-L504
Before it, we were sometimes encoding illegal (= empty) ACK frames, but that had the side effect of updating various bits of state...

**Comment by mxinden** [comment]
The panic stack-trace is happening across various versions, all the way back to Firefox 115.

![image](https://github.com/user-attachments/assets/7c93c5c4-2f60-4b4f-a652-ef4c1fc9029b)


https://crash-stats.mozilla.org/signature/?product=Firefox&signature=neqo_common%3A%3Acodec%3A%3AEncoder%3A%3Aencode_uint%3CT%3E&date=%3E%3D2024-09-17T11%3A46%3A00.000Z&date=%3C2024-09-24T11%3A46%3A00.000Z&_columns=date&_columns=product&_columns=version&_columns=build_id&_columns=platform&_columns=reason&_columns=address&_columns=install_time&_columns=startup_crash&_sort=-date&page=1#summary


This indicates  (doesn't proof) that it is a bug introduced before https://github.com/mozilla/neqo/commit/16e90850e2a038f3c14f67ee53556739b0d8f603.

**Comment by larseggert** [comment]
Glad it wasn't me :-)

**Comment by mxinden** [comment]
With my limited familiarity with crash-stats.mozilla.org in mind, that is my understanding too. That said, I have not been able to come up with a hypothesis why this would happen on one, but not any of the other operating systems.

**Comment by yjugl** [comment]
Crash-stats shows 99% correlation with aswJsFlt.dll. This is most likely a crash introduced by Avast DLL injection. Please work with :gstoll to ping our contacts and/or move the bug to the component External software affecting Firefox.

**Comment by mxinden** [comment]
Impressive find. Thank you @yjugl. I reached out to gstoll. Will report here in case there is progress.

**Comment by mxinden** [comment]
Next step is talking to Avast. See https://bugzilla.mozilla.org/show_bug.cgi?id=1919678#c4.

**Comment by yjugl** [comment]
Great, hopefully your analysis will help them figure out what they're doing that could make us reach this code path.
